### PR TITLE
Add URL Hash for Selected Tab

### LINKED ISSUE #22: Keep track of selected tab with url hashes
when we visit https://sce.sjsu.edu/transit/, the ui defaults to the first tab.
image

if we visit a different tab and refresh, the tab resets to the first. it would be nice to have the url include a hash when a tab is selected, so when we refresh the webpage we stay on the given tab.

information about url hashes can be found below:
https://developer.mozilla.org/en-US/docs/Web/API/URL/hash#examples

### game plan
- [ ] in [frontend/src/TransitPredictionsPage.js:36](https://github.com/SCE-Development/SCEta/blob/a6eef07a840b9325930ad2e746498e410e577ac3/frontend/src/TransitPredictionsPage.js#L36), we need to change the logic to use the hash of the url, if it exists to set the selected tab. like `#10th-%26-taylor` (`%26` is url encoded `&`) would highlight the `10th & Taylor` tab
- [ ] create a helper function that takes a selected tab that will set the hash of the url to the url encoded selected tab, with spaces replaced with dashes, for example `Santa Clara & 6th` becomes `Santa Clara-%26-6th`. the helper function also calls `setSelectedStop` so the tab changes in addition to the url.
- [ ] the helper function should be called in all locations in the above file where every `onChange and `onClick` prop is defined. instead of calling `setSelectedStop` directly

### PR DESCRIPTION
**Pull request by steeevin88**
selecting a tab adds a URL hash s.t when we refresh the webpage we stay on the given tab

resolves #22

### CODE REVIEW

**Review by evanugarte — COMMENTED** [review]
can we encode the urls in a nicer way? for example
```
Berryessa%20%2F%20North%20San%20Jose%20BART
```

```
berryessa-north-san-jose-bart
```

where any non alphanumerical characters become replaced with a `-`. we would need a mapping in javascript to map a hash to the correct tab

stackoverflow does this conversion, check out the url https://stackoverflow.com/questions/63749718/more-detail-about-ffmpeg-command

File: frontend/src/TransitPredictionsPage.js [hunk 1]
```diff
@@ -24,23 +24,46 @@ export default function TransitPredictionsPage() {
     getSCEtaPredictions();
   }, []);
 
+  function encode(stopName) {
+    return stopName.replace(/\s+/g, '-')
```
Comment (evanugarte) [inline frontend/src/TransitPredictionsPage.js:33]: add a comment above on what this replace is doing here and an example like
```js
// this replaces all non alpha numeric chars in the stop name with dashes.
// it also replaces the character "&" with "and", as well as convert all the
// characters in the stop to lowercase.
// for example "Santa Clara & 6th" becomes "santa-clara-and-6th"
```

File: frontend/src/TransitPredictionsPage.js [hunk 2]
```diff
@@ -79,29 +102,29 @@ export default function TransitPredictionsPage() {
         </div>
         {/* Tabs for larger screens */}
         <div className="hidden md:flex flex-row justify-center space-x-4 overflow-x-auto">
-          {stopOptions.map((stopName) => (
-            <button key={stopName} className={`px-4 py-2 text-xl font-semibold border-b-2 
+          {stopOptions.map((stopName, index) => (
+            <button key={index} className={`px-4 py-2 text-xl font-semibold border-b-2
```
Comment (evanugarte) [inline frontend/src/TransitPredictionsPage.js]: we cant use the stopName as the key like it was already?

https://stackoverflow.com/a/46735689

File: frontend/src/TransitPredictionsPage.js [hunk 3]
```diff
@@ -79,29 +102,29 @@ export default function TransitPredictionsPage() {
...
               {stopName}
             </button>
           ))}
         </div>
         <div>
-          {!!busData.length && busData.map((stop) => (
+          {!!busData.length && busData.map((stop, index) => (
             stop.name === selectedStop &&
-              <div key={stop.route} className="flex flex-col mt-4 p-6 min-w-80 max-h-[55vh] text-xl overflow-y-auto
+              <div key={index} className="flex flex-col mt-4 p-6 min-w-80 max-h-[55vh] text-xl overflow-y-auto
```
Comment (evanugarte) [inline frontend/src/TransitPredictionsPage.js]: use something else as the key, see my other comment as an explanation for why

File: frontend/src/TransitPredictionsPage.js [hunk 4]
```diff
@@ -79,29 +102,29 @@ export default function TransitPredictionsPage() {
...
                 bg-white rounded-lg border border-gray-200 shadow-md dark:bg-gray-800 dark:border-gray-700">
-                <div className="font-bold text-4xl mb-10">{stop.name}</div>
+                <div key={stop.name} className="font-bold text-4xl mb-10">{stop.name}</div>
```
Comment (evanugarte) [inline frontend/src/TransitPredictionsPage.js]: i think we only need the key in the parent component?

**Review by evanugarte — APPROVED** [review]

File: frontend/src/TransitPredictionsPage.js (hunk 4 above)
Comment (steeevin88) [reply frontend/src/TransitPredictionsPage.js]: resolved

File: frontend/src/TransitPredictionsPage.js (hunk 3 above)
Comment (steeevin88) [reply frontend/src/TransitPredictionsPage.js]: fixed; I use stop.name now

resolved

File: frontend/src/TransitPredictionsPage.js (hunk 2 above)
Comment (steeevin88) [reply frontend/src/TransitPredictionsPage.js]: yeah, I was getting an error in the console but I think this wasn't the issue

resolved

File: frontend/src/TransitPredictionsPage.js (hunk 1 above)
Comment (steeevin88) [reply frontend/src/TransitPredictionsPage.js:33]: added + resolved

**Pull request merged**